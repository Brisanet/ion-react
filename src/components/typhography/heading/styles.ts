import stitches from '../../../stitches.config';
const { styled } = stitches;

const defaultSettingHeading = {
  color: {
    primary: { color: '$neutral8' },
    secondary: { color: '$neutral7' },
    'dark-primary': { color: '$neutral1' },
    'dark-secondary': { color: '$neutral3' },
  },
  weight: {
    medium: { fontWeight: '400' },
    bold: { fontWeight: '600' },
  },
  spacingReset: { margin: 0, padding: 0 },
};

export const H1 = styled('h1', {
  ...defaultSettingHeading.spacingReset,
  variants: {
    color: defaultSettingHeading.color,
    size: {
      small: { fontSize: '24px', lineHeight: '32px' },
      medium: { fontSize: '32px', lineHeight: '40px' },
      normal: { fontSize: '40px', lineHeight: '48px' },
    },
    weight: defaultSettingHeading.weight,
  },
});

export const H2 = styled('h2', {
  ...defaultSettingHeading.spacingReset,
  variants: {
    color: defaultSettingHeading.color,
    size: {
      small: { fontSize: '20px', lineHeight: '28px' },
      medium: { fontSize: '24px', lineHeight: '32px' },
      normal: { fontSize: '32px', lineHeight: '40px' },
    },
    weight: defaultSettingHeading.weight,
  },
});

export const H3 = styled('h3', {
  ...defaultSettingHeading.spacingReset,
  variants: {
    color: defaultSettingHeading.color,
    size: {
      small: { fontSize: '16px', lineHeight: '24px' },
      medium: { fontSize: '20px', lineHeight: '28px' },
      normal: { fontSize: '24px', lineHeight: '32px' },
    },
    weight: defaultSettingHeading.weight,
  },
});

export const H4 = styled('h4', {
  ...defaultSettingHeading.spacingReset,
  variants: {
    color: defaultSettingHeading.color,
    size: {
      small: { lineHeight: '24px' },
      medium: { fontSize: '16px', lineHeight: '24px' },
      normal: { fontSize: '20px', lineHeight: '28px' },
    },
    weight: defaultSettingHeading.weight,
  },
});
