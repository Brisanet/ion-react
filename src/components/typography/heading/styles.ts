import { css, DefaultTheme, RuleSet, styled } from 'styled-components';
import {
  IonHeadingWeight,
  IonTypographyColorScheme,
  IonTypographySize,
  IonTypographyWeight,
} from '../typography.types';
import { defaultColor } from '../utils-typography';

type HeadingStyleProps = {
  $size: IonTypographySize;
  $color: IonTypographyColorScheme;
  $weight: IonHeadingWeight<IonTypographyWeight>;
};

const defaultHeading = (
  colorScheme: IonTypographyColorScheme,
  weight: HeadingStyleProps['$weight']
) => css`
  margin: 0;
  padding: 0;
  color: ${defaultColor('heading')[colorScheme].color};
  font-weight: ${{
    medium: 400,
    bold: 600,
  }[weight]};
`;

const defaultSizes: (
  theme: DefaultTheme,
  headingLevel: 1 | 2 | 3 | 4,
  size: IonTypographySize
) => RuleSet<object> = ({ font }, headingLevel, size) =>
  ({
    1: {
      small: css`
        ${font.size[24]}
      `,
      medium: css`
        ${font.size[32]}
      `,
      normal: css`
        ${font.size[40]}
      `,
    },
    2: {
      small: css`
        ${font.size[20]}
      `,
      medium: css`
        ${font.size[24]}
      `,
      normal: css`
        ${font.size[32]}
      `,
    },
    3: {
      small: css`
        ${font.size[16]}
      `,
      medium: css`
        ${font.size[20]}
      `,
      normal: css`
        ${font.size[24]}
      `,
    },
    4: {
      small: css`
        ${font.size[16]}
      `,
      medium: css`
        ${font.size[20]}
      `,
      normal: css`
        ${font.size[24]}
      `,
    },
  }[headingLevel][size]);

export const H1 = styled.h1<HeadingStyleProps>`
  ${({ theme, $size, $color, $weight }) => css`
    ${defaultHeading($color, $weight)}
    ${defaultSizes(theme, 1, $size)}
  `}
`;

export const H2 = styled.h2<HeadingStyleProps>`
  ${({ theme, $size, $color, $weight }) => css`
    ${defaultHeading($color, $weight)}
    ${defaultSizes(theme, 2, $size)}
  `}
`;

export const H3 = styled.h3<HeadingStyleProps>`
  ${({ theme, $size, $color, $weight }) => css`
    ${defaultHeading($color, $weight)}
    ${defaultSizes(theme, 3, $size)}
  `}
`;

export const H4 = styled.h4<HeadingStyleProps>`
  ${({ theme, $size, $color, $weight }) => css`
    ${defaultHeading($color, $weight)}
    ${defaultSizes(theme, 4, $size)}
  `}
`;
