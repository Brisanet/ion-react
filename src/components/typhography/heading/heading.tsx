import React from 'react';
import { IonHeadingInterface } from '../typhography.types';
import * as S from './styles';

const IonHeading: React.FC<IonHeadingInterface> = ({
  label,
  type,
  size = 'normal',
  colorScheme = 'primary',
  weight = 'medium',
}: IonHeadingInterface) => {
  const testid = `ion-heading-${type}`;
  const defaultProps = {
    'data-testid': testid,
    color: colorScheme,
    size,
    weight,
  };
  const headings = {
    h1: <S.H1 {...defaultProps}>{label}</S.H1>,
    h2: <S.H2 {...defaultProps}>{label}</S.H2>,
    h3: <S.H3 {...defaultProps}>{label}</S.H3>,
    h4: <S.H4 {...defaultProps}>{label}</S.H4>,
  };
  return headings[type];
};

export default IonHeading;
