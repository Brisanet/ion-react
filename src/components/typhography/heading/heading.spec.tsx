import React from 'react';
import { render, screen } from '@testing-library/react';
import { IonHeading, IonHeadingProps } from './heading';

const defaultProps: IonHeadingProps = { label: 'Ion React', type: 'h1' };

const sut = (props: IonHeadingProps = defaultProps) => {
  return render(<IonHeading {...props} />);
};

describe('<IonHeading />', () => {
  it('should render the heading with the correct label', () => {
    sut();
    expect(
      screen.getByRole('heading', { name: defaultProps.label })
    ).toBeInTheDocument();
  });

  it('should render the heading with the correct size', () => {
    sut();
    expect(screen.getByTestId('ion-heading-h1')).toHaveStyle({
      'font-size': '2em',
    });
    sut({ ...defaultProps, type: 'h2' });
    expect(screen.getByTestId('ion-heading-h2')).toHaveStyle({
      'font-size': '1.5em',
    });
  });

  it('should render the heading with the correct color', () => {
    sut();
    expect(screen.getByTestId('ion-heading-h1')).toHaveStyle({
      color: '$neutral8',
    });
    sut({ ...defaultProps, type: 'h2', colorScheme: 'secondary' });
    expect(screen.getByTestId('ion-heading-h1')).toHaveStyle({
      color: '$neutral7',
    });
  });

  it('should render the heading with the correct weight', () => {
    sut();
    expect(screen.getByTestId('ion-heading-h1')).toHaveStyle({
      'font-weight': 'bold',
    });
  });

  describe('snapshot tests', () => {
    it('should render heading h1', () => {
      const { container } = sut({ ...defaultProps, type: 'h1' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render heading h2', () => {
      const { container } = sut({ ...defaultProps, type: 'h2' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render heading h3', () => {
      const { container } = sut({ ...defaultProps, type: 'h3' });
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render heading h4', () => {
      const { container } = sut({ ...defaultProps, type: 'h4' });
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
