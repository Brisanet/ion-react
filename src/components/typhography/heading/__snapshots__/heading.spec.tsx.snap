// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IonHeading /> snapshot tests should render heading h1 1`] = `
<h1
  class="c-fXXVzC c-fXXVzC-ipNeOn-color-primary c-fXXVzC-dAfpSa-size-normal c-fXXVzC-cdWBIM-weight-medium"
  data-testid="ion-heading-h1"
>
  Ion React
</h1>
`;

exports[`<IonHeading /> snapshot tests should render heading h2 1`] = `
<h2
  class="c-fXXVzC c-fXXVzC-ipNeOn-color-primary c-fXXVzC-fJNGCD-size-normal c-fXXVzC-cdWBIM-weight-medium"
  data-testid="ion-heading-h2"
>
  Ion React
</h2>
`;

exports[`<IonHeading /> snapshot tests should render heading h3 1`] = `
<h3
  class="c-fXXVzC c-fXXVzC-ipNeOn-color-primary c-fXXVzC-fhGAOp-size-normal c-fXXVzC-cdWBIM-weight-medium"
  data-testid="ion-heading-h3"
>
  Ion React
</h3>
`;

exports[`<IonHeading /> snapshot tests should render heading h4 1`] = `
<h4
  class="c-fXXVzC c-fXXVzC-ipNeOn-color-primary c-fXXVzC-cQlbiW-size-normal c-fXXVzC-cdWBIM-weight-medium"
  data-testid="ion-heading-h4"
>
  Ion React
</h4>
`;
